File: C++/Baekjoon/Problem9019.cpp
#include <iostream>
#include <queue>

using namespace std;
int a, b;
bool visited[10000];

void bfs() {
    queue<pair<int, string>> q;
    q.push({a, ""});
    visited[a] = true;

    while (!q.empty()) {
        int tmp;
        int x = q.front().first;
        string op = q.front().second;
        q.pop();

        if (x == b) {
            cout << op << '\n';
            return;
        }

        for (int i = 0; i < 4; i++) {
            if (i == 0) { // D
                tmp = (2 * x) % 10000;
                if (!visited[tmp]) {
                    visited[tmp] = true;
                    q.push({tmp, op + "D"});
                }
            } else if (i == 1) { // S
                tmp = x - 1;
                if (tmp == -1) tmp = 9999;
                if (!visited[tmp]) {
                    visited[tmp] = true;
                    q.push({tmp, op + "S"});
                }
            } else if (i == 2) { // L
                tmp = (x % 1000) * 10 + x / 1000;
                if (!visited[tmp]) {
                    visited[tmp] = true;
                    q.push({tmp, op + "L"});
                }
            } else { // R
                tmp = (x % 10) * 1000 + x / 10;
                if (!visited[tmp]) {
                    visited[tmp] = true;
                    q.push({tmp, op + "R"});
                }
            }
        }
    }
}

int main() {
    int n;
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> a >> b;
        bfs();
        fill_n(visited, 10000, false);
    }
}

File: Python/Baekjoon/Problem9019.py
from collections import deque

def bfs(x):
    queue = deque([x,""])
    visited[x]=True

    while queue:
        x = queue.popleft()
        op = queue.popleft()

        if x == b :
            print(op,end='\n')
            return

        D = (x*2)%10000
        if visited[D] is False:
            visited[D]=True
            queue.append(D)
            queue.append(op+'D')

        S = x-1
        if S == -1 : S=9999
        if visited[S] is False:
            visited[S]=True
            queue.append(S)
            queue.append(op+'S')

        L = (x%1000)*10 + (x//1000)
        if visited[L] is False:
            visited[L]=True
            queue.append(L)
            queue.append(op+'L')

        R = (x%10)*1000 + (x//10)
        if visited[R] is False:
            visited[R]=True
            queue.append(R)
            queue.append(op+'R')

n = int(input())
for i in range(n) :
    visited = [False]*10000
    a,b = map(int,input().split())
    bfs(a)
